### PR TITLE
refactor: add a cache implementation for downloads

### LINKED ISSUE #197: Cache dir / volume
We should support an optional cache dir /volume to cache downloads, so we can mount a host volume / docker volume and reuse downloads to speedup later runs.

So Cache will be enabled if `BUILDPACK_CACHE` points to an existing folder.

### PR DESCRIPTION
**Pull request by Chumper**
Extracts the constants into its own file
Adds a cache implementation that can be used to cache downloaded files

Partially closes #197, at least for new tools that will use the functions

### CODE REVIEW

File: src/usr/local/buildpack/utils/cache.sh [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+# will attempt to download the file at the given url and stores it in the cache.
+# If this file already exists, will return the cached version
+# The cache will only used if BUILDPACK_CACHE_DIR is set
+# First argument is the url, second one is the filename (optional)
+function get_from_url () {
+  local url=${1}
+  check url true
+
+  local name
+  name=${2:-$(basename "${url}")}
```
Comment (viceice) [inline src/usr/local/buildpack/utils/cache.sh:12]: Thisd works for urls? 😏

File: src/usr/local/buildpack/utils/cache.sh [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+  if [ -n "${BUILDPACK_CACHE_DIR}" ] && [ -e "${BUILDPACK_CACHE_DIR}/${name}" ]; then
```
Comment (viceice) [inline src/usr/local/buildpack/utils/cache.sh:14]: We need to make sure the files are unique, can we use `TOOL_NAME` and `TOOL_VERSION` to create some implicit folders?

File: src/usr/local/buildpack/utils/cache.sh [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+# If the cache is not enabled it will download it to a temp folder
+# The second argument will be the filename if given
+function download_file () {
+  local url=${1}
+  check url true
+
+  local name
+  name=${2:-$(basename "${url}")}
+
+  local temp_folder=${BUILDPACK_CACHE_DIR:-$(mktemp -u)}
```
Comment (viceice) [inline src/usr/local/buildpack/utils/cache.sh:34]: We need to cleanup this when not using cache, as we don't want to have the downloads for prebuild images (eg python or node)

**Review by viceice — COMMENTED** [review]
Looks good so far

File: src/usr/local/buildpack/utils/cache.sh (hunk 1 above)
Comment (Chumper) [reply src/usr/local/buildpack/utils/cache.sh:12]: yes, urls are not that different than a file path and for our use case this is more than enough

File: src/usr/local/buildpack/utils/cache.sh (hunk 2 above)
Comment (Chumper) [reply src/usr/local/buildpack/utils/cache.sh:14]: I would like not to rely on the tool name and the tool version in this functions. 
Even if we agree that this is currently the only use case we should find a different solution.

The filename is not unique enough for you, so what about if we add the hash of the url to the name?
I would expect the full url to be unique enough without introducing the dependency on the tool name and version.

File: src/usr/local/buildpack/utils/cache.sh (hunk 3 above)
Comment (Chumper) [reply src/usr/local/buildpack/utils/cache.sh:34]: We do, my initial thought was to return the path and the user is responsible for cleanup.
We can introduce a general cache dir that we do not need to clean up like `/opt/buildpack/cache/` for downloaded tools when the cache is disabled. How does that sound?

File: src/usr/local/buildpack/utils/cache.sh (hunk 2 above)
Comment (viceice) [reply src/usr/local/buildpack/utils/cache.sh:14]: hash would be ok for me, as filename isn't unique enough.

File: src/usr/local/buildpack/utils/cache.sh (hunk 3 above)
Comment (viceice) [reply src/usr/local/buildpack/utils/cache.sh:34]: Currently we use `/tmp` which is cleaned up by `install-tool` and others. So can we use that as cache when no explicit caching is enabled?

Comment (Chumper) [reply src/usr/local/buildpack/utils/cache.sh:34]: `/tmp` is fine with me for now. 
We still need to have the discussion if we want to move the install folder alltogether but that can be done at a later stage.

**Review by viceice — APPROVED** [review]

**Pull request merged**